File: frontend/src/lib/ai/tools/filler-analysis.ts
/**
 * 필러워드 분석 도구
 * Python backend/langgraph/tools/filler_analysis.py 포팅
 */

export interface FillerAnalysisResult {
  fillerCount: number;
  fillerPercentage: number;
  totalWords: number;
  fillersByType: {
    sound: number;
    word: number;
    phrase: number;
  };
  fillersDetected: string[];
  mostCommonFillers: Array<[string, number]>;
  assessment: 'excellent' | 'acceptable' | 'excessive' | 'major_issue';
  recommendation: string;
  score: string; // A-D 등급
}

// 필러워드 패턴 (한국어)
const FILLER_PATTERNS = {
  // 소리 필러
  sounds: [
    /어+[\.…~]*/g,
    /음+[\.…~]*/g,
    /아+[\.…~]*/g,
    /에+[\.…~]*/g,
    /으+[\.…~]*/g,
  ],
  // 단어 필러
  words: [
    /\b그\b/g,
    /\b저\b/g,
    /\b뭐\b/g,
    /\b이제\b/g,
    /\b약간\b/g,
    /\b좀\b/g,
    /\b막\b/g,
    /\b진짜\b/g,
  ],
  // 구문 필러
  phrases: [
    /그러니까/g,
    /말하자면/g,
    /어떻게 보면/g,
    /솔직히/g,
    /사실/g,
    /뭐랄까/g,
    /그게/g,
    /있잖아/g,
  ],
};

export function analyzeFillers(transcript: string): FillerAnalysisResult {
  const words = transcript.trim().split(/\s+/).filter(Boolean);
  const totalWords = words.length;

  const fillersByType = { sound: 0, word: 0, phrase: 0 };
  const fillersFound: string[] = [];
  const fillerCounts: Record<string, number> = {};

  // 소리 필러 검출
  for (const pattern of FILLER_PATTERNS.sounds) {
    const matches = transcript.match(pattern) || [];
    fillersByType.sound += matches.length;
    matches.forEach((m) => {
      fillersFound.push(m);
      fillerCounts[m] = (fillerCounts[m] || 0) + 1;
    });
  }

  // 단어 필러 검출
  for (const pattern of FILLER_PATTERNS.words) {
    const matches = transcript.match(pattern) || [];
    fillersByType.word += matches.length;
    matches.forEach((m) => {
      fillersFound.push(m);
      fillerCounts[m] = (fillerCounts[m] || 0) + 1;
    });
  }

  // 구문 필러 검출
  for (const pattern of FILLER_PATTERNS.phrases) {
    const matches = transcript.match(pattern) || [];
    fillersByType.phrase += matches.length;
    matches.forEach((m) => {
      fillersFound.push(m);
      fillerCounts[m] = (fillerCounts[m] || 0) + 1;
    });
  }

  const fillerCount = fillersByType.sound + fillersByType.word + fillersByType.phrase;
  const fillerPercentage = totalWords > 0 ? (fillerCount / totalWords) * 100 : 0;

  // 가장 많이 사용된 필러
  const mostCommonFillers = Object.entries(fillerCounts)
    .sort(([, a], [, b]) => b - a)
    .slice(0, 5) as Array<[string, number]>;

  // 평가
  let assessment: FillerAnalysisResult['assessment'];
  let recommendation: string;
  let score: string;

  if (fillerPercentage <= 2) {
    assessment = 'excellent';
    recommendation = '필러워드 사용이 매우 적습니다. 훌륭합니다!';
    score = 'A';
  } else if (fillerPercentage <= 4) {
    assessment = 'acceptable';
    recommendation = '필러워드 사용이 적절한 수준입니다. 조금 더 줄이면 더 좋습니다.';
    score = 'B+';
  } else if (fillerPercentage <= 6) {
    assessment = 'excessive';
    const common = mostCommonFillers[0]?.[0] || '음';
    recommendation = `필러워드가 다소 많습니다 (${fillerPercentage.toFixed(1)}%). 특히 "${common}"을 자주 사용하시네요. 잠시 멈추는 연습을 해보세요.`;
    score = 'C+';
  } else {
    assessment = 'major_issue';
    const common = mostCommonFillers[0]?.[0] || '음';
    recommendation = `필러워드가 많아 전달력이 떨어집니다 (${fillerPercentage.toFixed(1)}%). "${common}" 대신 잠시 멈추는 습관을 들여보세요. 녹음하고 다시 들어보는 것도 도움이 됩니다.`;
    score = 'C';
  }

  return {
    fillerCount,
    fillerPercentage,
    totalWords,
    fillersByType,
    fillersDetected: fillersFound.slice(0, 10), // 최대 10개만 반환
    mostCommonFillers,
    assessment,
    recommendation,
    score,
  };
}

File: backend/langgraph/tools/filler_analysis.py
"""
필러워드(Filler Words) 분석 도구

"어...", "음...", "그..." 같은 필러워드를 감지하고 분석합니다.
"""

import re
from typing import List, Dict
from collections import Counter


# 한국어 필러워드 패턴 정의
FILLER_PATTERNS = {
    "sound": [
        r'\b어+[\.…~]*\b',
        r'\b음+[\.…~]*\b',
        r'\b아+[\.…~]*\b',
        r'\b에+[\.…~]*\b',
    ],
    "word": [
        r'\b그+\b',
        r'\b저+\b',
        r'\b뭐+\b',
        r'\b이제\b',
        r'\b약간\b',
        r'\b좀\b',
    ],
    "phrase": [
        r'그러니까',
        r'말하자면',
        r'어떻게 보면',
        r'솔직히',
        r'사실',
    ],
}


def analyze_fillers(transcript: str) -> dict:
    """필러워드 분석"""
    
    total_words = len(transcript.split())
    
    if total_words == 0:
        return {
            "filler_count": 0,
            "filler_percentage": 0.0,
            "fillers_by_type": {},
            "fillers_detected": [],
            "assessment": "excellent",
            "recommendation": "분석할 텍스트가 없습니다.",
        }
    
    fillers_by_type = {}
    all_fillers = []
    
    for filler_type, patterns in FILLER_PATTERNS.items():
        type_fillers = []
        for pattern in patterns:
            matches = re.findall(pattern, transcript, re.IGNORECASE)
            type_fillers.extend(matches)
        if type_fillers:
            fillers_by_type[filler_type] = type_fillers
            all_fillers.extend(type_fillers)
    
    filler_count = len(all_fillers)
    filler_percentage = round((filler_count / total_words) * 100, 1)
    
    assessment, recommendation = evaluate_fillers(filler_percentage, fillers_by_type)
    filler_counter = Counter(all_fillers)
    
    return {
        "filler_count": filler_count,
        "filler_percentage": filler_percentage,
        "total_words": total_words,
        "fillers_by_type": {k: len(v) for k, v in fillers_by_type.items()},
        "fillers_detected": all_fillers[:10],
        "most_common_fillers": filler_counter.most_common(5),
        "assessment": assessment,
        "recommendation": recommendation,
    }


def evaluate_fillers(percentage: float, by_type: Dict[str, List[str]]) -> tuple[str, str]:
    """필러워드 비율 기반 평가"""
    
    type_advice = {
        "sound": "'어...', '음...' 대신 잠시 멈춤(pause)을 사용해보세요.",
        "word": "'그', '뭐' 같은 습관적 단어를 의식하고 줄여보세요.",
        "phrase": "시간 벌기용 표현 대신 핵심을 바로 말해보세요.",
    }
    
    dominant_type = max(by_type.keys(), key=lambda k: len(by_type[k])) if by_type else None
    
    if percentage <= 2:
        return ("excellent", "필러워드 사용이 매우 적습니다.")
    elif percentage <= 4:
        base = "필러워드 사용이 허용 범위 내입니다."
        if dominant_type:
            return ("acceptable", f"{base} {type_advice.get(dominant_type, '')}")
        return ("acceptable", base)
    elif percentage <= 6:
        base = "필러워드가 다소 많습니다."
        if dominant_type:
            return ("excessive", f"{base} {type_advice.get(dominant_type, '')}")
        return ("excessive", base)
    else:
        return ("excessive", "필러워드가 매우 많습니다. '어...' 대신 의도적인 멈춤을 사용해보세요.")


def get_filler_score(percentage: float) -> str:
    """필러워드 비율을 점수로 변환"""
    if percentage <= 1:
        return "A"
    elif percentage <= 2:
        return "B+"
    elif percentage <= 3:
        return "B"
    elif percentage <= 4:
        return "C+"
    elif percentage <= 5:
        return "C"
    else:
        return "D"
